src/bugpoint.rs

```diff
@@ -547,6 +552,97 @@ impl Mutator for RemoveUnusedEntities {
     }
 }
 
+struct MergeBlocks {
+    ebb: Ebb,
+    prev_ebb: Option<Ebb>,
+}
+
+impl MergeBlocks {
+    fn new(func: &Function) -> Self {
+        Self {
+            ebb: func.layout.entry_block().unwrap(),
+            prev_ebb: None,
+        }
+    }
+}
+
+impl Mutator for MergeBlocks {
+    fn name(&self) -> &'static str {
+        "merge blocks"
+    }
+
+    fn mutation_count(&self, func: &Function) -> usize {
+        // N ebbs may result in at most N-1 merges.
+        ebb_count(func) - 1
+    }
+
+    fn mutate(&mut self, mut func: Function) -> Option<(Function, String, ProgressStatus)> {
+        let ebb = match func.layout.next_ebb(self.ebb) {
+            Some(ebb) => ebb,
+            None => return None,
+        };
+
+        self.ebb = ebb;
+
+        let mut cfg = ControlFlowGraph::new();
+        cfg.compute(&func);
+
+        if cfg.pred_iter(ebb).count() != 1 {
+            return Some((
+                func,
+                format!("did nothing for {}", ebb),
+                ProgressStatus::Skip,
+            ));
+        }
+
+        let pred = cfg.pred_iter(ebb).next().unwrap();
+
+        assert!(func.dfg.ebb_params(ebb).len() == func.dfg.inst_variable_args(pred.inst).len());
+
+        // If there were any EBB parameters in ebb, then the last instruction in pred will
+        // fill these parameters. Make the EBB params aliases of the terminator arguments.
+        for (ebb_param, arg) in func
+            .dfg
+            .detach_ebb_params(ebb)
+            .as_slice(&func.dfg.value_lists)
+            .iter()
+            .cloned()
+            .zip(func.dfg.inst_variable_args(pred.inst).iter().cloned())
+            .collect::<Vec<_>>()
+        {
+            if ebb_param != arg {
+                func.dfg.change_to_alias(ebb_param, arg);
+            }
+        }
+
+        // Remove the terminator branch to the current EBB.
+        func.layout.remove_inst(pred.inst);
+
+        // Move all the instructions to the predecessor.
+        while let Some(inst) = func.layout.first_inst(ebb) {
+            func.layout.remove_inst(inst);
+            func.layout.append_inst(inst, pred.ebb);
+        }
```

Comment:
Can you test this with basic-blocks flag, we have debug assertions which are preventing us from generating blocks which have a conditional branch followed by anything else than a jump.

If this fails with basic-blocks, please add a check that the previous instruction is not a branch, and if so skip this merge.

Reply:
I'm pretty sure there's no risk here: basic blocks structure is stable across the mutation. If blocks were respecting the basic blocks structure before, they'll still be afterwards, the branch instruction position being conserved when it's moved from the ebb to its predecessor.

Reply:
What about:

```
    brz
    jump ebb1

ebb1:
    iconst
    trap
```

That would result in:

```
    brz
    iconst
    trap
```

which us not allowed with basic blocks.

Reply:
Duh, I was very wrong :). Thanks.

So my understanding is that it becomes a verifier error and not a crash anymore if such a mutation is performed. In which case, it's just an invalid mutation and it'll be discarded. Is there anything else that needs to be done here?

Reply:
I think you are right. You may want to check for that case in the mutator, because otherwise there would be time wasted verifying the whole function.

Reply:
Yeah, it's nicer indeed. Done.